mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java

```diff
@@ -186,6 +200,16 @@ public void update(CRUDOperationContext ctx,
                 measure.begin("ctx.addDocument");
                 DocTranslator.TranslatedDoc translatedDoc=translator.toJson(document);
                 DocCtx doc=new DocCtx(translatedDoc.doc,translatedDoc.rmd);
+                if (memoryMonitor != null) {
+                    // if memory threshold is exceeded, this will throw an Error
+                    memoryMonitor.apply(doc);
```

Comment:
It's not even new to mongo, technically, since the only atomicity guarantee is with single documents; updating a range of documents may fail mongo-side mid operation as well.

On a separate note, I think there are other places in this algorithm the document is copied: at least when update starts, we copy the document again, and possibly also when the update is projected it may copy it as well (though I didn't read through that code well enough to see if it was just a pass-through-view upon the original or if a proper copy, though there is definitely non-zero memory overhead in either case). Should we monitor those too or is this approximation enough, given we have a separate threshold for writes?

The main variable is projections I guess, so I guess it ultimately depends if the projection makes another copy. If it does not, then the approximation is maybe fine: registering another copy is linearly proportional so it makes little difference other than how closely the threshold matches actual bytes used in memory. That said, if we tune that based on heap configuration, it may be less surprising if this matches more closely. I suppose we could just multiply the size calculated by * ~2 instead of adding the same doc to the monitor twice (i.e. use `(doc) -> JsonUtils.size(doc.getRoot() * 1.5 /* nice comment here explaining this is because we read the doc in memory as original, then copy it to do the update */ )`).

I used 1.5 in that example because I realized, when copying, it is not exactly * 2: primitives are reused since they are immutable (see ValueNode#deepCopy). Ultimately, not sure what is best without looking at real numbers. We know it is >1 * size but I don't know how much greater. Also, I think technically this may also impact hook calculation (double counting "copied-but-not-really" primitives). Maybe it's fine to be off by some, as long as it's documented as such when tuning thresholds. It's already an approximation to start, anyway.